drivers/usb/udc/udc_it82xx2.c

```diff
@@ -0,0 +1,1481 @@
+
+	struct k_fifo fifo;
+	struct k_work_delayable suspended_work;
+
+	struct k_thread thread_data;
+	struct k_sem suspended_sem;
+
+	/* FIFO_1/2/3 ready status */
+	bool fifo_ready[SHARED_FIFO_NUM];
+
+	/* FIFO_1/2/3 semaphore */
+	struct k_sem fifo_sem[SHARED_FIFO_NUM];
+
+	/* Record if the previous transaction of endpoint0 is STALL */
+	bool stall_is_sent;
+};
+
+struct usb_it82xx2_config {
+	struct usb_it82xx2_regs *const base;
+	const struct pinctrl_dev_config *pcfg;
+	const struct usb_it8xxx2_wuc wuc;
+	uint8_t usb_irq;
+	uint8_t wu_irq;
+	struct udc_ep_config *ep_cfg_in;
+	struct udc_ep_config *ep_cfg_out;
+	void (*make_thread)(const struct device *dev);
+};
+
+enum it82xx2_ep_ctrl {
+	EP_IN_DIRECTION_SET,
+	EP_IOS_ENABLE,
+	EP_ENABLE,
+	EP_DATA_SEQ_1,
+	EP_DATA_SEQ_TOGGLE,
+	EP_READY_ENABLE,
+};
+
+static inline void ep_set_halt(const struct device *dev, const uint8_t ep_idx, const bool enable)
+{
+	const struct usb_it82xx2_config *config = dev->config;
+	struct usb_it82xx2_regs *const usb_regs = config->base;
+	struct it82xx2_usb_ep_regs *ep_regs = usb_regs->usb_ep_regs;
+	union epn0n1_extend_ctrl_reg *ext_ctrl;
+	uint8_t idx = (ep_idx - 4) / 2;
+
+	ext_ctrl = usb_regs->fifo_regs[EP_EXT_REGS_9X].ext_4_15.epn0n1_ext_ctrl;
+	if (IT8XXX2_IS_EXTEND_ENDPOINT(ep_idx)) {
+		if (ep_idx % 2) {
+			ext_ctrl[idx].fields.epn1_send_stall_bit = enable;
+		} else {
+			ext_ctrl[idx].fields.epn0_send_stall_bit = enable;
+		}
+	} else {
+		ep_regs[ep_idx].ep_ctrl.fields.send_stall_bit = enable;
+	}
+}
+
+static volatile void *it82xx2_get_ext_ctrl(const struct device *dev, const uint8_t ep_idx,
+					   const enum it82xx2_ep_ctrl ctrl)
+{
+	uint8_t idx;
+	const struct usb_it82xx2_config *config = dev->config;
+	struct usb_it82xx2_regs *const usb_regs = config->base;
+	union epn0n1_extend_ctrl_reg *epn0n1_ext_ctrl =
+		usb_regs->fifo_regs[EP_EXT_REGS_9X].ext_4_15.epn0n1_ext_ctrl;
+	struct epn_ext_ctrl_regs *ext_ctrl =
+		usb_regs->fifo_regs[EP_EXT_REGS_DX].ext_0_3.epn_ext_ctrl;
+
+	if (ctrl == EP_IN_DIRECTION_SET || ctrl == EP_ENABLE) {
+		idx = ((ep_idx - 4) % 3) + 1;
+		return &ext_ctrl[idx].epn_ext_ctrl1;
+	}
+
+	idx = (ep_idx - 4) / 2;
+	return &epn0n1_ext_ctrl[idx];
+}
+
+static int it82xx2_usb_extend_ep_ctrl(const struct device *dev, const uint8_t ep,
+				      const enum it82xx2_ep_ctrl ctrl, const bool enable)
+{
+	const struct usb_it82xx2_config *config = dev->config;
+	struct usb_it82xx2_regs *const usb_regs = config->base;
+	struct epn_ext_ctrl_regs *ext_ctrl =
+		usb_regs->fifo_regs[EP_EXT_REGS_DX].ext_0_3.epn_ext_ctrl;
+	union epn_extend_ctrl1_reg *epn_ext_ctrl1 = NULL;
+	union epn0n1_extend_ctrl_reg *epn0n1_ext_ctrl = NULL;
+	const uint8_t ep_idx = USB_EP_GET_IDX(ep);
+
+	if (ctrl == EP_IN_DIRECTION_SET || ctrl == EP_ENABLE) {
+		epn_ext_ctrl1 =
+			(union epn_extend_ctrl1_reg *)it82xx2_get_ext_ctrl(dev, ep_idx, ctrl);
+	} else {
+		epn0n1_ext_ctrl =
+			(union epn0n1_extend_ctrl_reg *)it82xx2_get_ext_ctrl(dev, ep_idx, ctrl);
```

Comment:
No idea why it needs to return volatile, but maybe you can make it a bit simpler
```diff
diff --git a/drivers/usb/udc/udc_it82xx2.c b/drivers/usb/udc/udc_it82xx2.c
index cc9b4a82946..bf3cd6c42c9 100644
--- a/drivers/usb/udc/udc_it82xx2.c
+++ b/drivers/usb/udc/udc_it82xx2.c
@@ -187,16 +187,14 @@ static int it82xx2_usb_extend_ep_ctrl(const struct device *dev, const uint8_t ep
        struct usb_it82xx2_regs *const usb_regs = config->base;
        struct epn_ext_ctrl_regs *ext_ctrl =
                usb_regs->fifo_regs[EP_EXT_REGS_DX].ext_0_3.epn_ext_ctrl;
-       union epn_extend_ctrl1_reg *epn_ext_ctrl1 = NULL;
-       union epn0n1_extend_ctrl_reg *epn0n1_ext_ctrl = NULL;
+       volatile union epn_extend_ctrl1_reg *epn_ext_ctrl1 = NULL;
+       volatile union epn0n1_extend_ctrl_reg *epn0n1_ext_ctrl = NULL;
        const uint8_t ep_idx = USB_EP_GET_IDX(ep);
 
        if (ctrl == EP_IN_DIRECTION_SET || ctrl == EP_ENABLE) {
-               epn_ext_ctrl1 =
-                       (union epn_extend_ctrl1_reg *)it82xx2_get_ext_ctrl(dev, ep_idx, ctrl);
+               epn_ext_ctrl1 = it82xx2_get_ext_ctrl(dev, ep_idx, ctrl);
        } else {
-               epn0n1_ext_ctrl =
-                       (union epn0n1_extend_ctrl_reg *)it82xx2_get_ext_ctrl(dev, ep_idx, ctrl);
+               epn0n1_ext_ctrl = it82xx2_get_ext_ctrl(dev, ep_idx, ctrl);
        }
 
        switch (ctrl) {
```

Reply:
Done